### DISCUSSION TITLE
DNFPayload couldn't deal with USB Mass storage install media in USB-ZIP mode.

### DISCUSSION THREAD

**Discussion by slayercat** [rhinstaller/anaconda#3594] [General]
With `CentOS 8.x`  we found `anaconda` will boot but couldn't install from a USB stick -- it complains about "Error setting up software".  Even through `DRACUT_REPODIR` (`/run/install/repo` ) is mounted.
 
![image](https://user-images.githubusercontent.com/1755442/132471027-5239cd6b-9aed-4c7a-9238-32a19b37d0df.png)

Making a USB-ZIP mode install will create a partition formated in FAT32 in  USB stick, copys all content of iso to the partition.  `UltraISO` takes care of this. And `UEFI` will boot from that partition.

A further dig founds that, in commit 8a3c1182b34fd1514cca28c5d025691d40557da9, authors removed the `class PackagePayload`. In the good old days, `_setup_cdrom_device()` will check the baby and  do a repo setup for us. however, in commit 1e00603a504fc6380d9d449d3e6f807eab229c54, this baby-sitter has been retired.

I'm suggesting to hire a new nanny to care with the `DRACUT_REPODIR`. Because it's really annoying. Especially CentOS 8  requires a huge offline iso file (8 Gib).

Thanks.

**Comment by jkonecny12** [comment]
Hello, if I'm not mistaken you are creating the USB stick with `Rufus` tool which unfortunately creates unsupported partitioning layout as I already explained [here](https://github.com/rhinstaller/anaconda/pull/3529#issuecomment-891727423). In short the problem is that your tool for creating the ISO will change the structure and does not use ISO9660 format which is part of the ISO.

We have a plan to solve this another way, however, as workaround you should be able to just use `inst.repo` instead of `inst.stage2` which would tell Anaconda to not care about the partitioning and use that source as the installation source. And yes, our idea is similar to what you are proposing.

This change happened because Anaconda switched from the old code to a new solution, where we are trying to avoid re-using mounts from Dracut. We did that to avoid pretty complex logic to decide where to mount what and in what circumstances. Now the sources have a strict logic which is easy to follow and the code is nicely separated. Unfortunately, we were not aware of tooling which changing the ISO structure.

**Reply by slayercat** [reply]
Thanks for your quick reply.

Yes. this tool (`UltraISO`) changes the structure of file by design. However, If we may suggest, How about change `device_tree.FindOpticalMedia` to `FindMountablePartitions` ?  We have not found a reason to restrict install media to a miserable dvd rom.


Thanks again.

**Reply by jkonecny12** [reply]
Hi @poncovka what do you think about the proposal. I have a feeling that we already thought about that but I can't remember why we don't want to do that.

**Reply by slayercat** [reply]
we have made a prototype for the idea, and made some test in an offline machine. see #3599 for details.

If you feel that patch is valuable, we'll fix that unittest failure.

Looking forward to your comments and opinions. Thankyou.

**Reply by jkonecny12** [reply]
Thanks for the prototype, we will definitely look on the solution. :+1:
